File: src/chapters/chapter-05/csharp/Services/UserProfileService.cs
using Microsoft.SemanticKernel;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.DependencyInjection;

using System.ComponentModel;
using System.Text.Json.Serialization;

using AdvancedAIShoppingAssistant.Models;

namespace AdvancedAIShoppingAssistant.Services;

public class UserProfileService
{
    private readonly Dictionary<string, UserProfile> _userProfiles = new();

    public UserProfileService()
    {
        // Initialize with 3 fictitious users.
        _userProfiles.Add("user1", new UserProfile
        {
            UserId = "user1",
            Name = "Alice",
            Email = "alice@example.com",
            BrandAffinity = "BrandA",
            Budget = 500,
            CategoryInterests = new List<string> { "Electronics", "Books" },
            LatestVisitedProducts = new List<string> { "Smartphone X", "Cloud Book" }
        });

        _userProfiles.Add("user2",new UserProfile
        {
            UserId = "user2",
            Name = "Bob",
            Email = "bob@example.com",
            BrandAffinity = "BrandB",
            Budget = 1000,
            CategoryInterests = new List<string> { "Fitness", "Electronics" },
            LatestVisitedProducts = new List<string> { "Treadmill", "Smartwatch" }
        });

        _userProfiles.Add("user3", new UserProfile
        {
            UserId = "user3",
            Name = "Charlie",
            Email = "charlie@example.com",
            BrandAffinity = "BrandC",
            Budget = 200,
            CategoryInterests = new List<string> { "Books", "Fitness" },
            LatestVisitedProducts = new List<string> { "Yoga Mat", "Resistance Bands" }
        });
    }

    public Dictionary<string, UserProfile> UserProfiles
    {
        get => _userProfiles;
    }
}

File: src/chapters/chapter-05/python/services/user_profile_service.py
from typing import Dict
from models.user_profile import UserProfile

class UserProfileService:
    def __init__(self):
        self._user_profiles: Dict[str, UserProfile] = {}
        self._user_profiles["user1"] = UserProfile(
            user_id="user1",
            name="Alice",
            email="alice@example.com",
            brand_affinity="BrandA",
            budget=500.0,
            category_interests=["Electronics", "Books"],
            latest_visited_products=["Smartphone X", "Cloud Book"],
        )
        self._user_profiles["user2"] = UserProfile(
            user_id="user2",
            name="Bob",
            email="bob@example.com",
            brand_affinity="BrandB",
            budget=1000.0,
            category_interests=["Fitness", "Electronics"],
            latest_visited_products=["Treadmill B", "Smartwatch A"],
        )
        self._user_profiles["user3"] = UserProfile(
            user_id="user3",
            name="Charlie",
            email="charlie@example.com",
            brand_affinity="BrandC",
            budget=200.0,
            category_interests=["Books", "Fitness"],
            latest_visited_products=["Yoga Mat D", "Resistance Bands C"],
        )

    def get_user_profile(self, user_id: str):
        return self._user_profiles.get(user_id)

    @property
    def user_profiles(self):
        return self._user_profiles

